WinInstaller/WinInstaller/fsutils.cpp

```diff
@@ -0,0 +1,49 @@
+#include "fsutils.h"
+#include "winapi.h"
+
+
+namespace mywininstaller
+{
+	namespace fsutils
+	{
+		using std::filesystem::path;
+
+
+		void createDirectoryAndParentsTransacted(Transaction<path>& transaction, const path& dirPath)
+		{
+			const path& root = dirPath.root_path();
+			path subpath = "";
+			bool biggerThanRoot = false;
+			for (const path& part : dirPath)
+			{
+				subpath /= part;
+				if (biggerThanRoot || subpath > root)
+				{
+					// Caching this so we don't have to do lexicographic comparison each time.
+					biggerThanRoot = true;
+
+					if (winapi::createDirectoryThrows(subpath.c_str(), nullptr, true))
+					{
+						transaction.addAction(
```

Comment:
this code is not exception safe, if any operation fails in either the operations involved in creating the parameters, (e.g, memory allocation, the copy operation of "object", the string creation") or addAction itself, the operation will not be reverted

this is an important point, rethink why RAII is needed